src/pinnable_mapped_file.cpp

```diff
@@ -309,31 +450,43 @@ pinnable_mapped_file& pinnable_mapped_file::operator=(pinnable_mapped_file&& o)
 pinnable_mapped_file::~pinnable_mapped_file() {
    if(_writable) {
       if(_non_file_mapped_mapping) { //in heap or locked mode
-         _file_mapped_region = bip::mapped_region(_file_mapping, bip::read_write);
          save_database_file();
 #ifndef _WIN32
          if(munmap(_non_file_mapped_mapping, _non_file_mapped_mapping_size))
-            std::cerr << "CHAINBASE: ERROR: unmapping failed: " << strerror(errno) << std::endl;
+            std::cerr << "CHAINBASE: ERROR: unmapping failed: " << strerror(errno) << '\n';
 #endif
+      } else {
+         if (_sharable) {
+            if(_file_mapped_region.flush(0, 0, false) == false)
+               std::cerr << "CHAINBASE: ERROR: syncing buffers failed" << '\n';
+         } else {
+            save_database_file(); // must be before `this` is removed from _instance_tracker
+            if (auto it = std::find(_instance_tracker.begin(), _instance_tracker.end(), this); it != _instance_tracker.end())
+               _instance_tracker.erase(it);
+            _file_mapped_region = bip::mapped_region();
+            set_mapped_file_db_dirty(false);
+         }
       }
-      else
-         if(_file_mapped_region.flush(0, 0, false) == false)
-            std::cerr << "CHAINBASE: ERROR: syncing buffers failed" << std::endl;
       set_mapped_file_db_dirty(false);
    }
 }
 
 void pinnable_mapped_file::set_mapped_file_db_dirty(bool dirty) {
+   assert(_writable);
+   if (_file_mapped_region.get_address() == nullptr)
+      _file_mapped_region = bip::mapped_region(_file_mapping, bip::read_write, 0, _db_size_multiple_requirement);
    *((char*)_file_mapped_region.get_address()+header_dirty_bit_offset) = dirty;
-   if(_file_mapped_region.flush(0, 0, false) == false)
-      std::cerr << "CHAINBASE: ERROR: syncing buffers failed" << std::endl;
+   if (_file_mapped_region.flush(0, 0, false) == false)
+      std::cerr << "CHAINBASE: ERROR: syncing buffers failed" << '\n';
 }
 
 std::istream& operator>>(std::istream& in, pinnable_mapped_file::map_mode& runtime) {
    std::string s;
    in >> s;
    if (s == "mapped")
       runtime = pinnable_mapped_file::map_mode::mapped;
+   else if (s == "mapped_private")
```

Comment:
Just looking around at other option names thinking about consistency.. prior to 5.0 I think the only option values with`-`/`_` were `eos-vm-jit`, `eos-vm-oc` etc

It looks like 5.0's `http-category-address` now adds options like `chain_ro,127.0.0.1:8080`

sooo.. unless we want to back walk those to be `chain-ro` etc I guess `mapped_private` is fine

Reply:
I'd rather not change that as I've already provided the writup that was integrated into the release notes.

Reply:
Hmm, that is unfortunate. Wish that would have been `chain-ro`.